docs/wrapper-layer.md

```diff
@@ -0,0 +1,53 @@
+
+![Ergonomic to Logical API](assets/ergonomic-logical.svg)
+
+For example, in the above diagram, a client uses the ergonomic Intl.NumberFormat API with options to enable compact notation in long form.  The wrapper layer performs a bit of logic to figure out that this combination of options requires two functional units: plural rules and compact decimal format.  It then delegates to those two functions in the logical API.
+
+## Key Requirements for the Wrapper Layer
+
+### Host Language
+
+Since the ergonomic API depends on the host language (e.g., Objective-C calling into OmnICU via an FFI), the wrapper layer should be written in each individual host language, mapping to the logical API exported by the OmnICU library.
+
+This allows different languages to choose different styles for passing options into the library; for example, a Builder pattern could be used in Java, and an options bag could be used in JavaScript.
+
+### Code Slicing
+
+The wrapper layer should be suitable for tools to perform automatic dead code elimination.  By resolving options at compile time, the set of logical APIs required for a certain application can be known in time to perform static analysis; unused logical APIs can then be removed, reducing code and data size.
+
+In languages that support function inlining, the wrapper layer should generally be inlined into the call site, such that dead code elimination can tailor the set of logical APIs to each individual call site.
+
+Rust has great support for dead code elimination.  In languages where tooling for dead code elimination is not as strong, this should be taken into consideration when designing the ergonomic API specific to that language.  The ergonomic API might need to have a more logical flavor in order for code slicing to work as intended.
+
+### Data Loading
+
+In accordance with [principles.md](principles.md), the core library should not perform I/O; instead, data should be passed via a dynamic DataProvider ([data-pipeline.md](data-pipeline.md)).
+
+How does this fit into the wrapper layer model?
+
+1. The Ergonomic API receives a Data Provider.
+2. The Wrapper Layer loads the data asynchronously based on the required functional units, and then passes the data to the Functional APIs.
+
+This model decouples the data provider from the core business logic; with calls to the data provider happening in the wrapper layer, the core library does not need to make the decision of sync vs. async, etc.
+
+### Limited Amount of Code
+
+One of the strengths of OmnICU is that the core business logic is written once and shared by many different environments.
+
+Since the wrapper layer needs to be implemented separately for each host language, as much business logic as possible should be moved into the code slicing layer.  The wrapper layer should transform from the ergonomic API to the logical API and load the required data for the logical API; nothing more.
```

Comment:
At this point maybe mention transpiling or IDL as potential ways to generate some of the binding logic?

Reply:
Added